File: ocaml/chap03/binomialHeap.ml
module type OrderedType =
sig
  type t

  val eq : t -> t -> bool
  val lt : t -> t -> bool
  val leq : t -> t -> bool
end


module type S =
sig
  module Ord : OrderedType

  type t

  val empty : t
  val isEmpty : t -> bool

  val insert : Ord.t * t -> t
  val merge : t * t -> t

  val findMin : t -> Ord.t
  val deleteMin : t -> t
end


exception EMPTY


module Make (Order : OrderedType) =
struct
  module Ord = Order

  type tree = N of Ord.t * tree list

  type t = (int * tree) list

  let empty = []

  let isEmpty = function
      [] -> true
    | _ -> false

  let root (_, N (x, _)) = x

  let rank (r, _) = r

  let link ((r, (N (x1, c1) as t1)), (_, (N (x2, c2) as t2))) =
    if Ord.leq x1 x2 then (r + 1, N (x1, t2 :: c1))
    else (r + 1, N (x2, t1 :: c2))

  let rec insTree (t, ts) =
    match ts with
      [] -> [t]
    | t' :: ts' -> if rank t < rank t' then t :: ts else insTree (link (t, t'), ts')

  let rec insert (x, ts) = insTree ((0, N (x, [])), ts)

  let rec merge = function
      (ts1, []) -> ts1
    | ([], ts2) -> ts2
    | ((t1 :: ts1' as ts1), (t2 :: ts2' as ts2)) ->
      if rank t1 < rank t2 then t1 :: merge (ts1', ts2)
      else if rank t2 < rank t1 then t2 :: merge (ts1, ts2')
      else insTree (link (t1, t2), merge (ts1', ts2'))

  let rec removeMinTree = function
      [] -> raise EMPTY
    | [t] -> (t, [])
    | t :: ts -> let (t', ts') = removeMinTree ts in
      if Ord.lt (root t) (root t') then (t, ts) else (t' , t :: ts')

  let findMin ts = let (t, _) = removeMinTree ts in root t

  let ranking =
    let rec inner c = function
        [] -> []
      | x :: xs -> (c, x) :: inner (c + 1) xs
    in inner 0

  let deleteMin ts = let ((r, N (_, ts1)), ts2) = removeMinTree ts in merge (ranking (List.rev ts1), ts2)
end

File: rust/src/chap03/binomial_heap.rs
use crate::{
    chap02::list::List,
    common::{heap::Heap, ordered::Ordered, stack::Stack},
};

#[derive(Clone)]
struct BinomialHeap<T>(List<Tree<T>>);

#[derive(Clone)]
struct Tree<T>(usize, T, List<Tree<T>>);

impl<T: Ordered + Clone> Heap<T> for BinomialHeap<T> {
    fn empty() -> Self {
        BinomialHeap(List::empty())
    }

    fn is_empty(&self) -> bool {
        self.0.is_empty()
    }

    fn insert(&self, x: T) -> Self {
        self.ins_tree(Tree(0, x, List::empty()))
    }

    fn merge(&self, other: &Self) -> Self {
        match (self.0.get(), other.0.get()) {
            (_, Err(_)) => self.clone(),
            (Err(_), _) => other.clone(),
            (Ok((t1, ts1)), Ok((t2, ts2))) => {
                if t1.rank() < t2.rank() {
                    BinomialHeap(BinomialHeap(ts1).merge(other).0.cons(t1))
                } else if t2.rank() < t1.rank() {
                    BinomialHeap(self.merge(&BinomialHeap(ts2)).0.cons(t2))
                } else {
                    BinomialHeap(ts1)
                        .merge(&BinomialHeap(ts2))
                        .ins_tree(t1.link(&t2))
                }
            }
        }
    }

    fn find_min(&self) -> Result<T, String> {
        Ok(self.remove_min_tree()?.0.root())
    }

    fn delete_min(&self) -> Result<Self, String> {
        let (Tree(_, _, ts1), ts2) = self.remove_min_tree()?;
        Ok(BinomialHeap(ts1.reverse()).merge(&ts2))
    }
}

impl<T: Ordered + Clone> BinomialHeap<T> {
    fn ins_tree(&self, t: Tree<T>) -> Self {
        if let Ok((t_, ts_)) = self.0.get() {
            if t.rank() < t_.rank() {
                BinomialHeap(self.0.cons(t))
            } else {
                BinomialHeap(ts_).ins_tree(t.link(&t_))
            }
        } else {
            BinomialHeap(List::empty().cons(t))
        }
    }

    fn remove_min_tree(&self) -> Result<(Tree<T>, Self), String> {
        let (t, ts) = self.0.get().map_err(|_| "empty heap.")?;

        match BinomialHeap(ts.clone()).remove_min_tree() {
            Ok((t_, ts_)) => {
                if t.root().leq(&t_.root()) {
                    Ok((t, BinomialHeap(ts)))
                } else {
                    Ok((t_, BinomialHeap(ts_.0.cons(t))))
                }
            }
            Err(_) => Ok((t, BinomialHeap::empty())),
        }
    }
}

impl<T: Ordered + Clone> Tree<T> {
    fn link(&self, other: &Self) -> Self {
        let Tree(r, x1, c1) = self;
        let Tree(_, x2, c2) = other;

        if x1.leq(x2) {
            Tree(r + 1, x1.clone(), c1.cons(other.clone()))
        } else {
            Tree(r + 1, x2.clone(), c2.cons(self.clone()))
        }
    }

    fn rank(&self) -> usize {
        self.0
    }

    fn root(&self) -> T {
        self.1.clone()
    }
}

mod test {
    use crate::common::{heap::Heap, ordered::Ordered};

    use super::BinomialHeap;

    #[derive(Clone)]
    struct U32(u32);
    impl Ordered for U32 {
        fn eq(&self, other: &Self) -> bool {
            self.0 == other.0
        }

        fn leq(&self, other: &Self) -> bool {
            self.0 <= other.0
        }

        fn lt(&self, other: &Self) -> bool {
            self.0 < other.0
        }
    }

    #[test]
    fn test() {
        let mut h = BinomialHeap::empty();
        for n in [1, 6, 4, 4, 8, 3, 6, 7, 9, 5, 4, 2, 3, 6, 4, 8, 4] {
            h = h.insert(U32(n));
        }

        let mut v = vec![];
        while !h.is_empty() {
            v.push(h.find_min().unwrap().0);
            h = h.delete_min().unwrap();
        }
        println!("{:?}", v);
    }

    #[test]
    fn test_merge() {
        let h1 = {
            let mut h = BinomialHeap::empty();
            for n in [1, 6, 4, 4, 8, 3, 6, 7, 9, 5, 4, 2, 3, 6, 4, 8, 4] {
                h = h.insert(U32(n));
            }
            h
        };

        let h2 = {
            let mut h = BinomialHeap::empty();
            for n in [7, 2, 5, 6, 8, 4, 2, 7, 3, 3, 2, 7, 2, 7, 3, 9, 8, 3, 5] {
                h = h.insert(U32(n));
            }
            h
        };

        let mut h = h1.merge(&h2);

        let mut v = vec![];
        while !h.is_empty() {
            v.push(h.find_min().unwrap().0);
            h = h.delete_min().unwrap();
        }
        println!("{:?}", v);
    }
}
